### PR TITLE
0.9 ImageExporter, VideoExporter / VideoEncoder (based on FFMPEG)

### PR DESCRIPTION
**Pull request by lanice**
This adds:
- The ported ImageExporter
- A VideoEncoder based on FFMPEG
- A VideoExporter which stores the output of a RenderSurface into a video (using the VideoEncoder)
- Screenshotdialog and Videodialog in the gloperate-viewer to use those classes

It's not completely finished yet, especially the handling of the FFMPEG dependency. At the moment the CMakeScript sets the -DUSE_FFMPEG flag and the CPP sources have to handle the rest themselves.

### CODE REVIEW

File: source/examples/CMakeLists.txt [hunk 1]
```diff
@@ -13,3 +15,8 @@ endif()
 add_subdirectory(gloperate-glfw-example)
 add_subdirectory(gloperate-qt-example)
 add_subdirectory(gloperate-qtquick-example)
+
```
Comment (sbusch42) [inline source/examples/CMakeLists.txt]: Should not be necessary (the subproject checks for the dependency)

File: source/examples/gloperate-videotool-example/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+# 
+# Executable name and options
+# 
+
+# Target name
+set(target gloperate-videotool-example)
+
+# Exit here if required dependencies are not met
+if (NOT GLFW_FOUND)
```
Comment (sbusch42) [inline source/examples/gloperate-videotool-example/CMakeLists.txt]: Check for FFMPEG dependency?

File: source/gloperate-ffmpeg/include/gloperate-ffmpeg/VideoEncoder.h [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+namespace gloperate_ffmpeg
+{
+
+
+/**
+*  @brief
+*    Class for encoding single frames into a video using FFMPEG
+*/
+class GLOPERATE_FFMPEG_API VideoEncoder : public globjects::Referenced
```
Comment (sbusch42) [inline source/gloperate-ffmpeg/include/gloperate-ffmpeg/VideoEncoder.h]: Why globjects::Referenced? Does it even hold any OpenGL objects?

File: source/gloperate/include/gloperate/tools/AbstractVideoExporter.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+public:
+
+    /**
+    *  @brief
+    *    Destructor
+    */
+    virtual ~AbstractVideoExporter();
+
+    /**
+    *  @brief
```
Comment (sbusch42) [inline source/gloperate/include/gloperate/tools/AbstractVideoExporter.h]: No empty comments, please

File: source/gloperate/include/gloperate/tools/AbstractVideoExporter.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    */
+    virtual void init(const std::string & filename, gloperate::RenderSurface * surface, uint width, uint height, uint fps, uint length) = 0;
+
+    /**
```
Comment (sbusch42) [inline source/gloperate/include/gloperate/tools/AbstractVideoExporter.h]: No empty comments, please

File: source/gloperate/include/gloperate/tools/ImageExporter.h [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+
+/**
+*  @brief
+*    Tool to export images (screenshots) from a painter
+*/
+class GLOPERATE_API ImageExporter
+{
+public:
+    ImageExporter(RenderSurface * surface);
```
Comment (sbusch42) [inline source/gloperate/include/gloperate/tools/ImageExporter.h]: Comment

File: source/gloperate/include/gloperate/viewer/RenderSurface.h [hunk 7]
```diff
@@ -69,7 +71,24 @@ class GLOPERATE_API RenderSurface : public Surface
     */
     void setRenderStage(Stage * stage);
 
+    /**
+    *  @brief
```
Comment (sbusch42) [inline source/gloperate/include/gloperate/viewer/RenderSurface.h]: No empty comments, please

File: source/gloperate/source/tools/ImageExporter.cpp [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+#include <globjects/Framebuffer.h>
+
+
+namespace gloperate
+{
+
+
+ImageExporter::ImageExporter(RenderSurface * surface)
+    : m_surface(surface)
```
Comment (sbusch42) [inline source/gloperate/source/tools/ImageExporter.cpp]: Style (no indentation)

File: source/gloperate/source/viewer/RenderSurface.cpp [hunk 9]
```diff
@@ -20,7 +22,18 @@ RenderSurface::RenderSurface(ViewerContext * viewerContext)
...
 , m_mouseDevice(new MouseDevice(m_viewerContext->inputManager(), "Render Surface"))
 , m_keyboardDevice(new KeyboardDevice(m_viewerContext->inputManager(), "Render Surface"))
+, m_video(nullptr)
+, m_imageExporter(nullptr)
+, m_requestVideo(false)
 {
+    addFunction("createVideo", this, &RenderSurface::createVideo);
+    addFunction("exportImage", this, &RenderSurface::exportImage);
+
+    if (m_viewerContext->scriptEnvironment())
```
Comment (sbusch42) [inline source/gloperate/source/viewer/RenderSurface.cpp]: Would be better to do that from the outside (e.g., ViewerContext in registerSurface?)

File: source/gloperate/source/viewer/RenderSurface.cpp [hunk 10]
```diff
@@ -103,6 +153,18 @@ void RenderSurface::onRender(globjects::Framebuffer * targetFBO)
 {
     cppassist::details() << "onRender()";
 
+    if (m_requestVideo)
```
Comment (sbusch42) [inline source/gloperate/source/viewer/RenderSurface.cpp]: Add a big // [TODO] with an explanation why this is necessary (and what would be better)

**Comment by lanice** [comment]
Made obsolete by #147 

closed.

**Pull request closed without merging**